File: Leetcode/07_two_sum.py
from typing import List


def twoSum(nums: List[int], target: int) -> List[int]:
    nums_map = {}  # 검색 속도를 높이기 위한 딕셔너리
    for idx, num in enumerate(nums):
        if target - num in nums_map:  # 딕셔너리에 존재하면 인덱스들을 리턴
            return [idx, nums_map[target - num]]
        nums_map[num] = idx  # 딕셔너리에 현재 값이 존재하지 않으므로 숫자와 인덱스 삽입


print(twoSum([2, 7, 11, 15], 9,))
print(twoSum([2, 5, 5, 11], 10,))
print(twoSum([3, 2, 4], 6,))

File: Leetcode/07_two_sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function (nums, target) {
  // 검색을 위해서 배열을 해쉬테이블로 변환
  const hash = {};
  for (let i = 0; i < nums.length; i++) {
    // 해쉬 테이블에 target - nums[i] 값이 존재 할 때 종료
    if (hash[target - nums[i]] !== undefined) {
      return [hash[target - nums[i]], i];
    }
    // 해쉬 테이블에 추가
    hash[nums[i]] = i;
  }
  return [];
};

console.log(twoSum([2, 7, 11, 15], 9));
console.log(twoSum([2, 5, 5, 11], 10));
console.log(twoSum([3, 2, 4], 6));
